src/main/java/org/spongepowered/api/event/cause/Cause.java

```diff
@@ -48,52 +52,211 @@
  * some blocks, but tracing this event would be too complicated and thus
  * may not be attempted.</p>
  */
-public class Cause {
+@SuppressWarnings("unchecked")
+public abstract class Cause {
```

Comment:
You keep the "cause" nomenclature for stuff like `getAllCauses`, yet the class name is Cause. Maybe it deserves a different name like CauseSet?

Also, I'm not really sure what the difference is between an absent cause and a present one that has no "elements".